### PR TITLE
Add deserializer benchmarks

### PR DESCRIPTION
**Pull request by benmccallum**
Adds benchmarks for deserialization implementations.
* From body:
  * Introspection query
  * Introspection query with random vars and operation name
* Variables deserialization from string

### CODE REVIEW

**Comment by benmccallum** [comment]
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18362
Intel Core i7-4790 CPU 3.60GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


```
|         Method |        Mean |     Error |    StdDev |  Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |------------:|----------:|----------:|-------:|--------:|-------:|------:|------:|----------:|
| NewtonsoftJson |    285.9 ns |   2.76 ns |   2.44 ns |   1.00 |    0.00 | 0.8583 |     - |     - |   3.51 KB |
| SystemTextJson | 37,008.5 ns | 216.49 ns | 191.92 ns | 129.45 |    1.42 | 0.5493 |     - |     - |    2.4 KB |


TODO: 
* Add one for the deserialization into `Inputs` method.

**Comment by benmccallum** [comment]
This one's even worse @sungam3r, though I haven't had much of a chance to think about it. Just quickly whipped it up yesterday and am sharing it now early.

File: tests/Samples.Server.Tests/BaseTest.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 using System;
 using GraphQL.Server.Common;
 using GraphQL.Server.Transports.AspNetCore.Common;
+using GraphQL.Server;
```
Comment (sungam3r) [inline tests/Samples.Server.Tests/BaseTest.cs]: ?

File: src/Benchmarks/Benchmarks.csproj [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    <Compile Include="..\..\tests\Samples.Server.Tests\Serializer.cs" Link="Shared\Serializer.cs" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <ProjectReference Include="..\Transports.AspNetCore.NewtonsoftJson\Transports.AspNetCore.NewtonsoftJson.csproj" />
+    <ProjectReference Include="..\Transports.AspNetCore.SystemTextJson\Transports.AspNetCore.SystemTextJson.csproj" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <Folder Include="Shared\" />
```
Comment (sungam3r) [inline src/Benchmarks/Benchmarks.csproj]: ?

File: src/Benchmarks/Benchmarks.csproj [hunk 3]
```diff
@@ -0,0 +1,32 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <RootNamespace>GraphQL.Server.Benchmarks</RootNamespace>
+    <IsPackable>false</IsPackable>
```
Comment (sungam3r) [inline src/Benchmarks/Benchmarks.csproj]:
```suggestion
```

**Comment by sungam3r** [comment]
I think that is because `JsonDocument.ParseValue(ref reader)`

**Comment by sungam3r** [comment]
You can avoid reading into `JsonDocument` by executing `reader.Read()` and then swith over current token type. See https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to#support-dictionary-with-non-string-key

I believe it should solve the problem.

**Comment by sungam3r** [comment]
@benmccallum I have already fixed this
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.17134.1304 (1803/April2018Update/Redstone4)
Intel Core i5-6200U CPU 2.30GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
Frequency=2343753 Hz, Resolution=426.6661 ns, Timer=TSC
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X86 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X86 RyuJIT


```
|            Method |      Mean |     Error |    StdDev |    Median | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |----------:|----------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
|    NewtonsoftJson | 14.961 us | 0.9463 us | 2.7604 us | 13.391 us |  1.00 |    0.00 | 4.2725 |     - |     - |   6.59 KB |
| SystemTextJsonOld |  9.307 us | 0.3272 us | 0.3060 us |  9.216 us |  0.62 |    0.10 | 0.9003 |     - |     - |   1.39 KB |
|    SystemTextJson |  4.155 us | 0.0426 us | 0.0398 us |  4.154 us |  0.27 |    0.04 | 0.7706 |     - |     - |   1.19 KB |

**Comment by sungam3r** [comment]
@benmccallum `JsonDocument` is rather expensive API because it builds in-memory "database" for the whole json document. `JsonDocument` is for **random access** for json DOM. It is not necessary for deserialization process.

File: src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        //public DeserializeFromJsonBodyBenchmark()
+        {
+            _nsjDeserializer = new NsjDeserializer(s => { });
+            _stjDeserializer = new StjDeserializer(s => { });
+
+            var gqlRequest = new GraphQLRequest { Query = SchemaIntrospection.IntrospectionQuery };
+            var gqlRequestJson = Serializer.ToJson(gqlRequest);
+            var httpContext = new DefaultHttpContext();
+            httpContext.Features.Set<IRequestBodyPipeFeature>(new RequestBodyPipeFeature(httpContext));
+            httpContext.Request.Body = new MemoryStream(Encoding.UTF8.GetBytes(gqlRequestJson));
```
Comment (sungam3r) [inline src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs]: Don't you need to reset the position of this stream before starting each test?

Comment (sungam3r) [reply src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs]: After switching to preview-1505:

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.17134.1304 (1803/April2018Update/Redstone4)
Intel Core i5-6200U CPU 2.30GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
Frequency=2343753 Hz, Resolution=426.6661 ns, Timer=TSC
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X86 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X86 RyuJIT


```
|         Method |        Mean |     Error |    StdDev |  Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |------------:|----------:|----------:|-------:|--------:|-------:|------:|------:|----------:|
| NewtonsoftJson |    358.4 ns |   7.39 ns |   6.92 ns |   1.00 |    0.00 | 2.1553 |     - |     - |   3.31 KB |
| SystemTextJson | 88,478.9 ns | 351.83 ns | 329.11 ns | 246.92 |    4.75 | 0.8545 |     - |     - |   1.45 KB |

So I think the problem is shared state ans stream position.

Comment (benmccallum) [reply src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs]: I thought that the GlobalSetup ran every benchmark, but I guess the class isn't instantiated every benchmark, so yea, probably need to reset the stream pos.

File: tests/Samples.Server.Tests/BaseTest.cs (hunk 1 above)
Comment (benmccallum) [reply tests/Samples.Server.Tests/BaseTest.cs]: I moved the `Serializer` up a namespace level so it made sense when shared with Benchmarks project.

File: src/Benchmarks/Benchmarks.csproj (hunk 2 above)
Comment (benmccallum) [reply src/Benchmarks/Benchmarks.csproj]: Yea not sure, VS added this. Doesn't seem to be needed. Removing.

File: src/Benchmarks/Benchmarks.csproj (hunk 3 above)
Comment (benmccallum) [reply src/Benchmarks/Benchmarks.csproj]: I did this as I saw it was there in the graphql-dotnet Benchmarks project. Makes sense to never let it be packable.

File: src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs (hunk 4 above)
Comment (benmccallum) [reply src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs]: Yep, needed resetting. Will be fixed in next push.

**Comment by benmccallum** [comment]
**Latest results**

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18362
Intel Core i7-4790 CPU 3.60GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  Job-RSIGUH : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT

InvocationCount=1  UnrollFactor=1  

```
|         Method |     Mean |    Error |    StdDev |   Median | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|---------:|----------:|---------:|------:|--------:|------:|------:|------:|----------:|
| NewtonsoftJson | 26.38 us | 1.734 us |  5.003 us | 24.85 us |  1.00 |    0.00 |     - |     - |     - |  18.63 KB |
| SystemTextJson | 87.39 us | 3.760 us | 10.356 us | 83.50 us |  3.41 |    0.69 |     - |     - |     - |   2.77 KB |

**Comment by benmccallum** [comment]
So the deserialization of inputs, directly affected by the work you did on the `ObjectDictionaryConverter` is sweet. Interestingly the deserialization from body is 4x slower on S.T.Json than NewtJson; though it does allocate 1/9th the memory, so that's a win. 

Maybe there's more we can do here though. 

I've got to keep going on other stuff at work, since I've spent so much time on these things already, but feel free to investigate if you have time and you can make changes in here too, np.

File: src/Benchmarks/Benchmarks.csproj (hunk 3 above)
Comment (sungam3r) [reply src/Benchmarks/Benchmarks.csproj]: graphql-dotnet has a bit different build settings - graphql-dotnet/graphql-dotnet#1481 , it packs all that is possible so we need to opt-out `IsPackable` explicitly.

**Comment by sungam3r** [comment]
The problem is somewhere in async I/O.

**Comment by sungam3r** [comment]
After #329 we can be sure that if there are performance problems, then entirely in this project.

**Comment by sungam3r** [comment]
@benmccallum I would like to merge this PR. It’s good in itself that you added benchmarks into the repo. Then we can experiment with the code and examine its performance. Could you resolve the conflicts and prepare the PR for the merge?

File: src/Benchmarks/SchemaIntrospection.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+";
+
```
Comment (sungam3r) [inline src/Benchmarks/SchemaIntrospection.cs]:
```suggestion
```

File: src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        [Benchmark(Baseline = true)]
+        public Task<GraphQLRequestDeserializationResult> NewtonsoftJson() => _nsjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest);
+
+        [Benchmark]
+        public Task<GraphQLRequestDeserializationResult> SystemTextJson() => _stjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest);
+
+        [Benchmark]
+        public Task<GraphQLRequestDeserializationResult> NewtonsoftJson_WithOpNameAndVariables() => _nsjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest);
```
Comment (sungam3r) [inline src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs]:
```suggestion
        public Task<GraphQLRequestDeserializationResult> NewtonsoftJson_WithOpNameAndVariables() => _nsjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest2);
```

File: src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        public Task<GraphQLRequestDeserializationResult> NewtonsoftJson() => _nsjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest);
+
+        [Benchmark]
+        public Task<GraphQLRequestDeserializationResult> SystemTextJson() => _stjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest);
+
+        [Benchmark]
+        public Task<GraphQLRequestDeserializationResult> NewtonsoftJson_WithOpNameAndVariables() => _nsjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest2);
+
+        [Benchmark]
+        public Task<GraphQLRequestDeserializationResult> SystemTextJson_WithOpNameAndVariables() => _stjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest);
```
Comment (sungam3r) [inline src/Benchmarks/Benchmarks/DeserializeFromJsonBodyBenchmark.cs]:
```suggestion
        public Task<GraphQLRequestDeserializationResult> SystemTextJson_WithOpNameAndVariables() => _stjDeserializer.DeserializeFromJsonBodyAsync(_httpRequest2);
```

**Pull request merged**